### ISSUE TITLE
Implementation of `UPDF_ReaderBPLibrary::PDF_Get_Texts` wrong

### ISSUE THREAD

**Issue by mara004** [FF-Projects-UE/PDF_Reader#4]
https://github.com/FF-Projects-UE/PDF_Reader/blob/7cf2b6209d9f4d1d7e341015aa818aa043f4f7d8/Source/PDF_Reader/Private/PDF_ReaderBPLibrary.cpp#L909-L946

See https://groups.google.com/g/pdfium/c/HwqzzGWWXVU/m/XCc8CODiAgAJ:
> a) The innermost for-loop seems to be gravely wrong.
> b) I don't think you should close the FPDF_PAGE before the FPDF_TEXTPAGE.
> c) A static size buffer is inflexible and may truncate text if not large enough.

**Comment by ffreality** [comment]
It was personally mine topic and we solved it after learning correct way to convert unsigned short pointer to FString. For allocation size, devs told that GetText doesn't return lenght like other functions. So, we need to "guess" its size. We agreed that each character should be 4 byte. So, size should be char count * 4.

currently we have two problems.
1- Loading external font breaks PDF file
2- Loading Image either doesn't show up or breaks PDF.

**Comment by mara004** [comment]
> We agreed that each character should be 4 byte. So, size should be char count * 4.

Mmh, no, we now think `count * 2` is sufficient, if you look at the [latest post](https://groups.google.com/g/pdfium/c/HwqzzGWWXVU/m/shunElL3AQAJ) from @jerbob92 on the mailing list.

**Comment by ffreality** [comment]
(Count + 1) * because null operator.
Actually I don't care about allocation size and don't follow their additional talks. If they say 2 is enough, okay but as I say before my want to give my focus to solve both external font and image loading.

**Comment by ffreality** [comment]
I looked their talks and they okay with "2" because pdfium "currently" doesn't support surrogate pair chars. Not because UTF16LE is only 2 bytes.

So, I won't change it. Because when PDFium solves this, I don't want to hassle with it again. Evantually we need to convert it 4, again.
On the other hand, this is not a low performance device targeted code. If anyone have a problem about memory when they try to import a PDF file, they shouldn't use Unreal at first place.

But if you really want to change it, feel free to create your own fork.

**Comment by mara004** [comment]
You misunderstood. This API (we think) *is designed to treat* a surrogate pair as two separate characters, and the current functions won't change. If pdfium fixes this, a new API providing the exact memory size would be added.

TL;DR allocate 2 BPC (plus null terminator)

**Comment by ffreality** [comment]
If you think so, okay I will do it today.
Do you have a suggestion about other problems ?
image, font and image extraction

**Comment by ffreality** [comment]
Also I want to change save function. Currently it uses global variable because C callbacks don't support C++ lambda captures.

I will create a save actor class with "FString Path UPROPERTY". When someone wants to save a file, they will just spawn an actor. This allows users save pdfs whenever they want
